File: WePeiYang/Setting/Controller/SettingViewController.m
//
//  SettingViewController.m
//  WePeiYang
//
//  Created by Qin Yubo on 15/11/10.
//  Copyright © 2015年 Qin Yubo. All rights reserved.
//

#import "SettingViewController.h"
#import "data.h"
#import "wpyCacheManager.h"
#import "GPATableViewController.h"
#import "MsgDisplay.h"
#import "wpyDeviceStatus.h"

@interface SettingViewController ()

@end

@implementation SettingViewController {
    NSArray *titleArr;
    NSArray *footerArr;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.tableView reloadData];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationController.view.backgroundColor = [UIColor whiteColor];
    self.clearsSelectionOnViewWillAppear = YES;
    
    titleArr = @[@"设置", @"关于"];
    footerArr = @[@"", [NSString stringWithFormat:@"微北洋 %@ Build %@\n2015 TWT Studio", [wpyDeviceStatus getAppVersion], [wpyDeviceStatus getAppBuild]]];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Private methods

- (void)clearGPACacheFinishBlock:(void(^)())block {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"清除" message:@"确定要清除 GPA 登录信息及缓存吗？" preferredStyle: UIAlertControllerStyleActionSheet];
    UIAlertAction *clearAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil) style:UIAlertActionStyleDestructive handler:^(UIAlertAction *action) {
        [wpyCacheManager removeCacheDataForKey:GPA_USER_NAME_CACHE];
        [wpyCacheManager removeCacheDataForKey:GPA_CACHE];
        [MsgDisplay showSuccessMsg:@"GPA 缓存及登录信息已清除"];
        block();
    }];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", nil) style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:clearAction];
    [alert addAction:cancel];
    [alert setModalPresentationStyle:UIModalPresentationPopover];
    alert.popoverPresentationController.permittedArrowDirections = 0;
    alert.popoverPresentationController.sourceView = self.view;
    alert.popoverPresentationController.sourceRect = self.view.frame;
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)loginGpaFinish:(void(^)())block {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"请输入天大办公网的账号密码" message:@"由于部分原因，暂时无法校验您账号密码是否正确，请保证输入准确无误。\n如反复要求您输入验证码，可能是您输入错误，请注销后重新登录。" preferredStyle:UIAlertControllerStyleAlert];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
        textField.placeholder = @"请输入办公网账号";
    }];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
        textField.placeholder = @"请输入办公网密码";
        textField.secureTextEntry = YES;
    }];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil) style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        NSDictionary *dic = @{@"username": alertController.textFields[0].text,
                              @"password": alertController.textFields[1].text};
        [wpyCacheManager saveCacheData:dic withKey:GPA_USER_NAME_CACHE];
        block();
    }];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", nil) style:UIAlertActionStyleCancel handler:nil];
    [alertController addAction:cancel];
    [alertController addAction:okAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return titleArr.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return 1;
            break;
            
        case 1:
            return 1;
            break;
            
        default:
            return 0;
            break;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"simpleTableViewIdentifier" forIndexPath:indexPath];
    if (indexPath.section == 0) {
        if (indexPath.row == 0) {
            cell.textLabel.text = [wpyCacheManager cacheDataExistsWithKey:GPA_USER_NAME_CACHE] ? @"注销 GPA 登录信息" : @"登录办公网";
        }
    } else if (indexPath.section == 1) {
        NSArray *arr = @[@"关于微北洋"];
        cell.textLabel.text = arr[indexPath.row];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return titleArr[section];
}

- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section {
    return footerArr[section];
}

#pragma mark - Table View delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSInteger section = indexPath.section;
    NSInteger row = indexPath.row;
    
    if (section == 0) {
        if (row == 0) {
            if ([wpyCacheManager cacheDataExistsWithKey:GPA_USER_NAME_CACHE]) {
                [self clearGPACacheFinishBlock:^{
                    [self.tableView reloadData];
                }];
            } else {
                [self loginGpaFinish:^{
                    [self.tableView reloadData];
                }];
            }
        }
    } else if (section == 1) {
        if (row == 0) {
            [self performSegueWithIdentifier:@"pushAbout" sender:nil];
        }
    }
    
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}

@end

File: WePeiYang/Setting/Controller/SettingViewController.swift
//
//  SettingViewController.swift
//  WePeiYang
//
//  Created by Qin Yubo on 16/1/11.
//  Copyright © 2016年 Qin Yubo. All rights reserved.
//

import UIKit
import LocalAuthentication
import BlocksKit
import CoreSpotlight

class SettingViewController: UITableViewController {
    
    // 用 map 来定义需要计算的全局变量
    var appVer: String {
        return wpyDeviceStatus.getAppVersion()
    }
    var appBuild: String {
        return wpyDeviceStatus.getAppBuild()
    }
    var titleArr = ["账号", "设置", "关于"]
    var footerArr: [String] {
        return ["", "允许 Spotlight 索引后，您可以在主屏幕搜索页里搜索您的成绩信息。", "微北洋 \(appVer) Build \(appBuild)"]
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
        
        self.clearsSelectionOnViewWillAppear = true
        self.title = "设置"
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.tableView.reloadData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Private methods
    
    private func loginOrLogout() {
        if AccountManager.tokenExists() {
            let alert = UIAlertController(title: "注销", message: "确定要注销吗？", preferredStyle: .ActionSheet)
            let okAction = UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .Destructive, handler: {action in
                AccountManager.removeToken()
                wpyCacheManager.removeCacheDataForKey(GPA_CACHE)
                wpyCacheManager.removeCacheDataForKey(GPA_USER_NAME_CACHE)
                MsgDisplay.showSuccessMsg("注销成功！")
                self.tableView.reloadData()
            })
            let cancelAction = UIAlertAction(title: NSLocalizedString("Cancel", comment: ""), style: .Cancel, handler: nil)
            alert.addAction(okAction)
            alert.addAction(cancelAction)
            alert.modalPresentationStyle = .Popover
            alert.popoverPresentationController?.permittedArrowDirections = UIPopoverArrowDirection.Any
            alert.popoverPresentationController?.sourceView = self.tableView.cellForRowAtIndexPath(NSIndexPath(forRow: 0, inSection: 0))
            alert.popoverPresentationController?.sourceRect = self.tableView.cellForRowAtIndexPath(NSIndexPath(forRow: 0, inSection: 0))!.bounds
            self.presentViewController(alert, animated: true, completion: nil)
            
        } else {
            let loginVC = LoginViewController(nibName: nil, bundle: nil)
            self.presentViewController(loginVC, animated: true, completion: nil)
        }
        
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 3
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            return 1
        case 1:
            return 2
        case 2:
            return 2
        default:
            return 0
        }
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = UITableViewCell(style: .Default, reuseIdentifier: "reuseIdentifier")
        let section = indexPath.section
        let row = indexPath.row
        switch section {
        case 0:
            switch row {
            case 0:
                cell.textLabel?.text = AccountManager.tokenExists() ? "注销" : "登录"
            default:
                break
            }
        case 1:
            switch row {
            case 0:
                cell.textLabel?.text = "使用 Touch ID"
                let TOUCH_ID_KEY = "touchIdEnabled"
                let touchIDSwitch = UISwitch()
                let defaults = NSUserDefaults()
                touchIDSwitch.on = defaults.boolForKey(TOUCH_ID_KEY)
                touchIDSwitch.bk_addEventHandler({handler in
                    if touchIDSwitch.on == false {
                        // 关闭touchID，需要验证
                        let authContext = LAContext()
                        var error: NSError?
                        guard authContext.canEvaluatePolicy(.DeviceOwnerAuthenticationWithBiometrics, error: &error) else {
                            MsgDisplay.showErrorMsg("您的设备不支持 Touch ID")
                            touchIDSwitch.setOn(false, animated: true)
                            defaults.setBool(false, forKey: TOUCH_ID_KEY)
                            return
                        }
                        authContext.evaluatePolicy(.DeviceOwnerAuthenticationWithBiometrics, localizedReason: "改变Touch ID相关设置须验证您的指纹", reply: {(success, error) in
                            if success {
                                defaults.setBool(false, forKey: TOUCH_ID_KEY)
                            } else {
                                // 注意：TouchID之后的UI操作放在主线程
                                dispatch_async(dispatch_get_main_queue(), {
                                    touchIDSwitch.setOn(true, animated: true)
                                })
                                defaults.setBool(true, forKey: TOUCH_ID_KEY)
                            }
                        })
                    } else {
                        // 打开touchID，不需要验证
                        let authContext = LAContext()
                        var error: NSError?
                        guard authContext.canEvaluatePolicy(.DeviceOwnerAuthenticationWithBiometrics, error: &error) else {
                            MsgDisplay.showErrorMsg("您的设备不支持 Touch ID")
                            touchIDSwitch.setOn(false, animated: true)
                            return
                        }
                        defaults.setBool(true, forKey: TOUCH_ID_KEY)
                    }
                }, forControlEvents: .ValueChanged)
                cell.accessoryView = touchIDSwitch
                cell.selectionStyle = .None
            case 1:
                cell.textLabel?.text = "允许 Spotlight 索引"
                let ALLOW_SPOTLIGHT_KEY = "allowSpotlightIndex"
                let spotlightSwitch = UISwitch()
                let defaults = NSUserDefaults()
                if defaults.objectForKey(ALLOW_SPOTLIGHT_KEY) == nil {
                    if #available(iOS 9.0, *) {
                        defaults.setBool(true, forKey: ALLOW_SPOTLIGHT_KEY)
                    } else {
                        defaults.setBool(false, forKey: ALLOW_SPOTLIGHT_KEY)
                    }
                }
                spotlightSwitch.on = defaults.boolForKey(ALLOW_SPOTLIGHT_KEY)
                spotlightSwitch.bk_addEventHandler({handler in
                    if #available(iOS 9.0, *) {
                        if spotlightSwitch.on == false {
                            // 关闭
                            CSSearchableIndex.defaultSearchableIndex().deleteAllSearchableItemsWithCompletionHandler({error in
                                if (error != nil) {
                                   print(error)
                                }
                            })
                            defaults.setBool(false, forKey: ALLOW_SPOTLIGHT_KEY)
                        } else {
                            defaults.setBool(true, forKey: ALLOW_SPOTLIGHT_KEY)
                        }
                    } else {
                        // Fallback on earlier versions
                        MsgDisplay.showErrorMsg("您的 iOS 版本不支持 Spotlight 索引\n请升级至最新版本")
                    }
                }, forControlEvents: .ValueChanged)
                cell.accessoryView = spotlightSwitch
                cell.selectionStyle = .None
            default:
                break
            }
        case 2:
            switch row {
            case 0:
                cell.textLabel?.text = "关于微北洋"
                cell.accessoryType = .DisclosureIndicator
            case 1:
                cell.textLabel?.text = "反馈"
                cell.accessoryType = .DisclosureIndicator
            default:
                break
            }
        default:
            break
        }
        return cell
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return titleArr[section]
    }
    
    override func tableView(tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        return footerArr[section]
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let section = indexPath.section
        let row = indexPath.row
        
        switch section {
        case 0:
            switch row {
            case 0:
                self.loginOrLogout()
            default:
                break
            }
        case 1:
            break
        case 2:
            switch row {
            case 0:
                break
            case 1:
                let feedbackVC = FeedbackViewController(style: .Grouped)
                self.navigationController?.showViewController(feedbackVC, sender: nil)
            default:
                break
            }
        default:
            break
        }
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
    }

    /*
    // Override to support conditional editing of the table view.
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }
    */

    /*
    // Override to support editing the table view.
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if editingStyle == .Delete {
            // Delete the row from the data source
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
        } else if editingStyle == .Insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    */

    /*
    // Override to support rearranging the table view.
    override func tableView(tableView: UITableView, moveRowAtIndexPath fromIndexPath: NSIndexPath, toIndexPath: NSIndexPath) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(tableView: UITableView, canMoveRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return false if you do not want the item to be re-orderable.
        return true
    }
    */

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
